### PR TITLE
[#77963] End running reservations on an instrument before starting a new one

### PR DESCRIPTION
**Pull request by jdzak**
https://pm.tablexi.com/issues/77963

### CODE REVIEW

File: app/support/products/scheduling_support.rb [hunk 1]
```diff
@@ -25,6 +25,10 @@ def admin_reservations
     self.reservations.admin
   end
 
+  def started_reservations
+    self.reservations.where("reservations.actual_start_at IS NOT NULL")
```
Comment (jdzak) [inline app/support/products/scheduling_support.rb]: Add `AND actual_end_at IS NULL`

File: app/models/reservation.rb [hunk 2]
```diff
@@ -115,6 +115,7 @@ def self.overlapping(start_at, end_at)
   #####
 
   def start_reservation!
+    product.started_reservations.each(&:end_reservation!)
```
Comment (jdzak) [inline app/models/reservation.rb]: Mark reservation as complete only. Don't set end time or price policy.

**Comment by jdzak** [comment]
@jhanggi Comments addressed

File: app/support/products/scheduling_support.rb [hunk 3]
```diff
@@ -25,6 +25,11 @@ def admin_reservations
     self.reservations.admin
   end
 
+  def started_reservations
+    self.purchased_reservations
+      .where("reservations.actual_start_at IS NOT NULL AND reservations.actual_end_at IS NULL")
```
Comment (jdzak) [inline app/support/products/scheduling_support.rb]: See if we can dry this up with existing scopes

File: spec/models/reservation_spec.rb [hunk 4]
```diff
@@ -664,6 +664,46 @@
...
+        order.state = 'validated'
+        order.purchase!
+      end
+
+      it 'completes the running reservation' do
+        reservation.start_reservation!
+        expect(running.reload).to be_complete
+      end
+
+      it 'sets the orders as a problem order' do
```
Comment (jhanggi) [inline spec/models/reservation_spec.rb]: I would probably add a test that it doesn't set the actual_end_at of the previously-running reservation.

**Comment by jhanggi** [comment]
Overall this looks pretty good. Just the one minor comment. Question, though: how are you handling some of the code changes that you're sharing across PRs? e.g. `complete!`

File: spec/models/reservation_spec.rb (hunk 4 above)
Comment (jdzak) [inline spec/models/reservation_spec.rb]: Sure.

**Comment by jdzak** [comment]
For `complete!`, I just duplicated the code across the PRs (including code placement) and was planning on letting git do the conflict resolution. Since the `complete!` code is the same, git should just merge the PR cleanly.

**Comment by jdzak** [comment]
@jhanggi comments addressed.

File: app/support/products/scheduling_support.rb [hunk 5]
```diff
@@ -25,6 +25,11 @@ def admin_reservations
     self.reservations.admin
   end
 
+  def started_reservations
+    self.purchased_reservations
+      .merge(Reservation.relay_in_progress)
```
Comment (jdzak) [inline app/support/products/scheduling_support.rb]: Should this be only getting the reservations with actual_start_at set (relays) or should using reserve_start_at?

**Comment by jhanggi** [comment]
:+1: Let's merge this one in, then we can rebase the other PRs since there's some overlap.

**Pull request closed without merging**